### PR TITLE
Try a different random port selection algorithm

### PR DESCRIPTION
**Pull request by tylerbenson**
Linux seems rather aggressive at reusing closed ports which leads to a race condition. Try a bigger range to reduce chance of immediate reuse.

### CODE REVIEW

File: dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/utils/PortUtils.java [hunk 1]
```diff
@@ -26,17 +28,29 @@ public static int randomOpenPort() {
   }
 
   /** Open up a random, server socket and keep it open. */
-  public static ServerSocket randomOpenSocket() {
-    try {
-      ServerSocket socket = new ServerSocket(0);
-      socket.setReuseAddress(true);
```
Comment (lpriima) [inline dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/utils/PortUtils.java:32]: I would say just remove this line and it should already help.

File: dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/utils/PortUtils.java [hunk 2]
```diff
@@ -26,17 +28,29 @@ public static int randomOpenPort() {
...
-      return socket;
-    } catch (final IOException ioe) {
-      ioe.printStackTrace();
-      return null;
+  public static ServerSocket randomOpenSocket() throws TimeoutException {
+    long timeout = TimeUnit.SECONDS.toNanos(1);
+    long startTime = System.nanoTime();
+    while (true) {
+      try {
+        return new ServerSocket(nextRandomPort());
```
Comment (lpriima) [inline dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/utils/PortUtils.java:36]: `new ServerSocket(0)` should already choose from "ephemeral port" range. If we are going to test Windows or some other OSs port above 5000 or 65535 might not work...

**Comment by richardstartin** [comment]
Isn't this papering over the cracks? How does this reduce the rate of test failure caused by port conflicts? To echo things which have been said before, if you request port zero, a free port is found and assigned. It is possible to obtain the port _somehow_ (some ways are cleaner than others) from each server instance. The impediment to doing this is the structure of the tests, which expect to be in control of the operating system's ports, to somehow reserve a port and later bind to it. Refactoring the tests (which is painful, because they were written in Groovy) to allow the operating system to choose the port, or to use containers, are possible solutions.

**Comment by jbachorik** [comment]
@richardstartin Indeed. All the attempts to outsmart the OS to pre-allocate ports I've seen during my career ended up in tears. In short - there is no reliable way to pre-allocate ports. Period.

**Comment by tylerbenson** [comment]
I understand that this is subverting the OS's ability to choose an open port.  You can see my other attempt at reducing this problem (as well as discussion that led to this PR) here: #2570.

The reason this approach was considered is that Linux seems very prone to reusing recently closed ports when given the opportunity to do so, which results in a race condition between the random port selection closing the port and it being reopened by the server.  The only guaranteed solution is to pass `0` into the server and see what port is actually selected. This is a much bigger task which is why I was trying to find a "bandaid" that can help in the interim.

**Comment by richardstartin** [comment]
But it's not at all clear that it would help, let alone how.

**Comment by richardstartin** [comment]
par exemple

```
javax.ws.rs.ProcessingException: Failed to start Grizzly HTTP server: Address already in use
	at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpServerFactory.createHttpServer(GrizzlyHttpServerFactory.java:297)
	at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpServerFactory.createHttpServer(GrizzlyHttpServerFactory.java:118)
	at GrizzlyTest.startServer(GrizzlyTest.groovy:36)
	at datadog.trace.agent.test.base.WithHttpServer.setupSpec(WithHttpServer.groovy:30)
Caused by: java.net.BindException: Address already in use
	at sun.nio.ch.Net.bind(Net.java:469)
	at sun.nio.ch.Net.bind(Net.java:458)
	at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:220)
	at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:88)
	at org.glassfish.grizzly.nio.transport.TCPNIOBindingHandler.bindToChannelAndAddress(TCPNIOBindingHandler.java:131)
	at org.glassfish.grizzly.nio.transport.TCPNIOBindingHandler.bind(TCPNIOBindingHandler.java:88)
	at org.glassfish.grizzly.nio.transport.TCPNIOTransport.bind(TCPNIOTransport.java:248)
	at org.glassfish.grizzly.nio.transport.TCPNIOTransport.bind(TCPNIOTransport.java:228)
	at org.glassfish.grizzly.nio.transport.TCPNIOTransport.bind(TCPNIOTransport.java:219)
	at org.glassfish.grizzly.http.server.NetworkListener.start(NetworkListener.java:709)
	at org.glassfish.grizzly.http.server.HttpServer.start(HttpServer.java:278)
	at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpServerFactory.createHttpServer(GrizzlyHttpServerFactory.java:294)
	... 3 more
```

https://app.circleci.com/pipelines/github/DataDog/dd-trace-java/7392/workflows/4776f0e2-2977-4816-8620-a301871cc68f/jobs/141177

File: dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/utils/PortUtils.java (hunk 1 above)
Comment (dougqh) [reply dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/utils/PortUtils.java:32]: I'm not sure that's true.  If Linux is aggressive reusing ports, then we're going to have a race.

Maybe we should test this in isolation from the server tests and prove it works before we make the change.

**Comment by dougqh** [comment]
I think a fall in the middle of this debate.

To be clear, I think the test should let the server find and use a port itself whenever possible.
The biggest failure here is the design of the test system not allowing for the server to the selection itself.  Certainly that needs to change -- although, I'm willing to have that be a separate PR.

It would be nice to know how many of the servers support this.  If it is almost all, then I'd say we should focus on that problem first.

If we have a non-trivial number of servers that don't allow for automatic port selection, then I think there might be something to a different port selection strategy.  But I think that depends on first proving that the default port selection strategy doesn't work for use case.

So before we commit such a change, I want to see evidence that it actually reduces the rate of port collisions.


As for retries (which is where this started), I think that's the wrong place to start.  If indeed the port selection algorithm aggressively reuses ports, then I don't think retry is going to help all that much.

So again, before we commit such a change, I want to see evidence that it actually reduces the rote of collisions.

**Comment by tylerbenson** [comment]
Going to close this PR for now then.

**Pull request closed without merging**